### PR TITLE
Add flag capability, and rewrite using more formal Alternative instance

### PR DESCRIPTION
**Pull request by nc6**

### CODE REVIEW

File: src/Options/Schema.hs [hunk 1]
```diff
@@ -65,10 +65,18 @@ data Option a = Option {
 instance Functor Option where
   fmap f (Option n d b) = Option n d (fmap f b)
 
-data Block a =
-    SingleArgument (Argument a)
-  | Subsection (Schema a)
+data Block a where
+  SingleArgument :: Argument a -> Block a
+  Subsection :: Schema a -> Block a
+  Flag :: Block Bool
```
Comment (qnikst) [inline src/Options/Schema.hs]: [minor, doc] some comments would be helpful.

File: src/Options/Schema/Applicative.hs [hunk 2]
```diff
@@ -41,7 +41,17 @@ mkBlockParser sn _ (Subsection a) = mkParser . (hoistAlt go) $ a
     fstName = case (head sn) of
       LongName x -> x
       ShortName x -> [x]
+mkBlockParser n d Flag = mkFlagParser n d 
 
+-- | Make a flag parser
+mkFlagParser :: [Name] -> Description -> Parser Bool
+mkFlagParser n d = let
+  names = foldl1 (<>) . map mkName $ n
```
Comment (qnikst) [inline src/Options/Schema/Applicative.hs]: [style] `names = mconcat $ mkName <$> n`?

File: src/Options/Schema/Applicative.hs [hunk 3]
```diff
@@ -41,7 +41,17 @@ mkBlockParser sn _ (Subsection a) = mkParser . (hoistAlt go) $ a
...
+  props = foldl1 (<>) $ names :
+    catMaybes [
+        fmap help $ dSummary d
+    ]
```
Comment (qnikst) [inline src/Options/Schema/Applicative.hs]: [style] I'm not sure but for me it's more nice to have:
```
maybe names (mconcat . (names:))  $  fmap help $ dSummary d
```
there are more smiles inside :))

File: src/Control/Alternative/Freer.hs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  fmap f (Ap x g) = x `Ap` fmap (f .) g
+  fmap _ Empty = Empty
+  fmap f (Or a b) = Or (f <$> a) (f <$> b)
+  fmap f (Some l) = f . (:[]) <$> l
+
+instance Functor f => Applicative (Alt f) where
+  pure = Pure
+  {-# INLINE pure #-}
+  (Pure f) <*> y = fmap f y
+  (Ap a f) <*> b = a `Ap` (flip <$> f <*> b)
```
Comment (qnikst) [inline src/Control/Alternative/Freer.hs]: [minor, style] I'd use `Ap a (flip <$> f <*> b)` to keep that in the same style as other equations.

File: src/Control/Alternative/Freer.hs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+hoistAlt f (Or a b) = Or (hoistAlt f a) (hoistAlt f b)
+hoistAlt f (Some a) = Some $ hoistAlt f a
+
+-- | Interpret the free alternative over f using the Alternative semantics for f.
+retractAlt :: Alternative f => Alt f a -> f a
+retractAlt (Pure a) = pure a
+retractAlt (a `Ap` f) = a <**> retractAlt f
+retractAlt Empty = empty
+retractAlt (Or a b) = retractAlt a <|> retractAlt b
+retractAlt (Some a) = some $ retractAlt a
```
Comment (qnikst) [inline src/Control/Alternative/Freer.hs]: [minor, style] newline :)

**Comment by qnikst** [comment]
Will add that I'd really love to see any value in a flag not just `Bool`.

**Comment by nc6** [comment]
Comments addressed :-)

**Pull request merged**